### ISSUE TITLE
prediction API for probabilistic classifiers

### ISSUE THREAD

**Issue by fkiraly** [alan-turing-institute/MLJ.jl#34]
I think we should *not* follow the common prediction interface for probabilistic classification which returns an array of probabilities. The reason is that this does not generalize well to multi-label classification, or regression. In fact, it does not generalize at all without various hacks and "secret conventions" that make these cases quite ugly in current interface designs such as mlr or sklearn.

I might be biased, but I'd prefer that probabilistic learners return a distribution interface such as in
https://github.com/alan-turing-institute/skpro
This avoids having to think about how exactly to represent distributions in a return object and proliferating a large set of arbitrary design conventions depending on the metadata/task.

**Comment by fkiraly** [comment]
From #33, moved answer here since more topical.

>>on a similar note: you want to think very carefully about what the (back-end and front-end) return type of “probabilistic + multivariate” should be…

>What do you suggest?

Predict (for methods with the"probabilistic" tag) returns a "distribution" struct on which you can call eval, e.g.,
eval(y_distr, class_vec)
with inputs y_distr = "M-vector of predictive pmf"
and class_vec = "M-vector of class labels" (categorical)

gives you a M-vector with predicted probability masses.

I'd think that is a much more stable design.

**Comment by fkiraly** [comment]
Just noticed that eval is already a quite central method in Julia.
Replace eval above by "pmf_eval" or similar.

**Comment by ablaom** [comment]
I strongly agree that predicting a distribution object would be ideal. However, there are some disadvantages:

- The alternative conventions you refer to are pretty well-entrenched in the main use cases (i.e. excluding multivariate targets)

- The return-distribution-object alternative is more abstract (less friendly to the unsophisticated user).

- We need to implement an extra interface to convert the distribution-type predictions into forms the loss functions expect, or write all our own loss functions (although we may have do this anyway, I haven't looked into this enough yet). What do you do in scipro?

- The barrier to entry for implementing the MLJInterface for new classifiers (in external packages) goes up

I understand that @vollmersj prefers the more conventional approach. Perhaps he will weight in on this discussion.

**Comment by fkiraly** [comment]
>The alternative conventions you refer to are pretty well-entrenched in the main use cases (i.e. excluding multivariate targets)

I think this is actually an advantage - we do it the better way from the start on.
The alternative conventions also lead to confusion, especially in third party interfacing. For example, the keras sklearn interface's predict behaves as predict_proba.

>The return-distribution-object alternative is more abstract (less friendly to the unsophisticated user).

Yes, but the unsophisticated user also gets confused when you return an array that has (# labels minus on) columns. What are the columns? And why is one missing?

>We need to implement an extra interface to convert the distribution-type predictions into forms the loss functions expect, or write all our own loss functions (although we may have do this anyway, I haven't looked into this enough yet). What do you do in scipro?

The loss functions just takes the distribution object - after all, they have to compare a distribution to a prediction.

Also, try to write the loss function for the cross-entropy loss aka log-loss when you have to make case distinctions in the "classical" interface. My educated guess is that you haven't done so, since otherwise you may see this as *supporting* my point?

>The barrier to entry for implementing the MLJInterface for new classifiers (in external packages) goes up

Not if the interface is well done.

As in the case of losses, you have to compare a solution to its alternatives - including how annoying it is in the classical case to try to find out in which format the probabilities are expected to be returned, where they go in the array, how many columns there are, and why on earth there is an interface difference between the case of binary classification and classification with 3 or more labels.

**Comment by fkiraly** [comment]
Hm, I might have had an APIphany, if you allow the pun.

There might be a way to have the cake and eat it:
by default, probabilistic classifiers could return not an array but a *data frame* with an attribute or decoration (can Julia do that?) that specifies it to be a vector of pmf. On the other hand, methods that expect a distribution and get a data frame will know it is the distribution object for a pmf.

I.e., we go with the distribution object API, but simply make the additional convention that a decorated data frame is the pmf object.

We could also make the convention that a data frame with column names "mean" and "var" (and decorated with the "normal" flag) is a vector of normal pdfs with that mean and variance - allowing us to get rid of the "variance" method for prediction.

More complicated methods (such as "compute the L2-norm" or "return cdf value") can be incrementally implemented via Julia's dispatch architecture quite nicely, i.e., don't need to be anticipated.

What do you think?

**Comment by fkiraly** [comment]
Just checked out
https://github.com/JuliaStats/Distributions.jl
which was suggested by @ablaom as an output type.

I think it has an excellent API and would hence endorse this. However, CI tests currently fail at Julia 0.7?

Also, at some later point, one may want to think carefully about an "array-of-distributions" or a "data-frame-of-distributions" type, and more specifically about an "outer-product-of-distribution" first-order type, which is the distribution representation of a tuple of independent random variables.

**Comment by ablaom** [comment]
**Update**

After recent meeting with @fkiraly, @ysimillides, @sjvollmer , @darenasc, it was agreed to go ahead with the following design:

- Supervised models should be separated into those that predict a probability distribution for each input pattern and those that make direct "point" predictions. We are calling these *probabilistic* models and *deterministic* models respectively, and the distinction is reflected in the model's `metadata`. 

The idea is that if the user wants a point-prediction from a probabilistic model, then they compose their model with an appropriate transformer. That said, the current API permits models to optionally implement a `predict_mode` (for classifiers) or a `predict_mean` (for regressors) that yield point-predictions for probabilistic models. Such methods might, for example, save memory or computation, but at the moment I would view this as "bells and whistles".

- For the present at least, "distributions" are either subtypes of `Distributions` in the `Distributions.jl` package referred to in the post above, or, where necessary, custom types sharing the same method interface. 

For more detail, refer to the latest version of [adding_new_models.md](https://github.com/alan-turing-institute/MLJ.jl/blob/master/doc/adding_new_models.md)

**Comment by ablaom** [comment]
Regarding the question of whether we need a special array-of-distributions type: I am hopeful that ordinary arrays whose element type has the distribution trait (that we will define) will suffice. Julia type dispatch should allow us to define things like outer products, etc.

**Comment by fkiraly** [comment]
Regarding the other bells and whistles, predict_mode, predict_mean: very much agreed! This turns the common sklearn/Weka API convention around, but makes much more sense. After all, it is not the *probabilistic* functionality that's "tacked on" but the *deterministic* one. So the API should reflect this - either by transformer fusion or subsidiary public methods.

**Comment by fkiraly** [comment]
Another subsidiary point: distributions need to implement the L2-norm in order to support computation of Brier loss (in classification case) and integrated squared loss (in regression case). Currently, this is neither a mandatory nor optional method in the Distributions.jl API.

**Comment by ablaom** [comment]
Closed in favour of more focused discussions.
